skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +174,341 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+                          normalized, unweighted):
+
+    # temporarily store u_counts and v_counts in a 2D array as that's what
+    # _vectorize_counts_and_tree takes
+    u_counts = np.asarray(u_counts)
+    v_counts = np.asarray(v_counts)
+    counts = np.vstack([u_counts, v_counts])
+    counts_by_node, tree_index, branch_lengths = \
+        _vectorize_counts_and_tree(counts, otu_ids, tree)
+    # unpack counts vectors for single pairwise UniFrac calculation
+    u_node_counts = counts_by_node[0]
+    v_node_counts = counts_by_node[1]
+
+    u_total_count = u_counts.sum()
+    v_total_count = v_counts.sum()
+
+    return (u_node_counts, v_node_counts, u_total_count, v_total_count,
+            tree_index)
+
+
+def _unweighted_unifrac(u_node_counts, v_node_counts, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        Unweighted UniFrac distance between samples.
+
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    unique = np.logical_xor(u_node_counts, v_node_counts)
+    observed = np.logical_or(u_node_counts, v_node_counts)
+    return (branch_lengths * unique).sum() / (branch_lengths * observed).sum()
+
+
+def _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                      v_total_count, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+        The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+        respectively. This could be computed internally, but since this is a
+        private method and the calling function has already generated these
+        values, this saves an iteration over each of these vectors.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        Weighted UniFrac distance between samples.
+
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    if u_total_count > 0:
+        # convert to relative abundances if there are any counts
+        u_ = u_node_counts / u_total_count
```

Comment:
Let's call ``u_`` and ``v_`` something more descriptive (e.g., ``u_node_proportions``) and then return them so they don't have to be recomputed if we're computing normalized weighted UniFrac.